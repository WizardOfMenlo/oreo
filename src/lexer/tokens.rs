//! Definition of all the tokens that we use

use super::error::LexicalError;
use serde::Serialize;

/// The main token type that we work on
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Token<'a> {
    /// Any of the keywords (apart from and and or)
    Keyword(Keyword),

    /// (, ), ;, ,
    Punctuation(Punctuation),

    /// Addition, mult, ...
    Operator(Operator),

    /// A constant
    Literal(Literal<'a>),

    /// An identifier
    Identifier(&'a str),

    /// A comment
    Comment(&'a str),

    /// Any kind of error we might encounter
    Error(LexicalError<'a>),
}

impl<'a> Token<'a> {
    /// Is the current token an error?
    pub fn is_error(&self) -> bool {
        match self {
            Token::Error(_) => true,
            _ => false,
        }
    }

    /// Is the current token a comment?
    pub fn is_comment(&self) -> bool {
        match self {
            Token::Comment(_) => true,
            _ => false,
        }
    }

    /// Is this token the same as an other (where Int(1) == Int(2))
    pub fn same_kind<'b>(&self, other: &Token<'b>) -> bool
    where
        'b: 'a,
    {
        // This should match every singleton
        if self == other {
            true
        } else {
            // Match literals
            match (self, other) {
                (Self::Literal(Literal::Integer(_)), Self::Literal(Literal::Integer(_))) => true,
                (Self::Literal(Literal::Boolean(_)), Self::Literal(Literal::Boolean(_))) => true,
                (Self::Literal(Literal::String(_)), Self::Literal(Literal::String(_))) => true,
                (Self::Identifier(_), Self::Identifier(_)) => true,
                (Self::Comment(_), Self::Comment(_)) => true,
                _ => false,
            }
        }
    }
}

/// The Keywords we have
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Keyword {
    Program,
    Begin,
    End,
    Var,
    Print,
    Println,
    Get,
    While,
    If,
    Then,
    Else,
    Procedure,
    Return,
}

/// Possible Punctuation marks
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Punctuation {
    Comma,
    Semicolon,
    BracketOpen,
    BracketClose,
}

/// All the operators (including boolean ones)
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Operator {
    Assignement,
    Plus,
    Minus,
    Times,
    Divide,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
    Or,
    And,
    Not,
}

/// Constants
#[allow(missing_docs)]
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub enum Literal<'a> {
    Integer(isize),
    Boolean(bool),
    String(&'a str),
}
