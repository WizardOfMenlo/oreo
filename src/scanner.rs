use crate::range::RangedObject;

#[derive(Debug, PartialEq, Clone)]
pub enum ScannedItem<'a> {
    Str(&'a str),
    UnclosedStr(&'a str),
    Comment(&'a str),
    UnclosedComment(&'a str),
    Rest(&'a str),
}

pub fn scan(input: &str) -> impl Iterator<Item = RangedObject<ScannedItem>> {
    LineScannerIt { input, pos: 0 }
}

pub struct LineScannerIt<'a> {
    input: &'a str,
    pos: usize,
}

impl<'a> Iterator for LineScannerIt<'a> {
    type Item = RangedObject<ScannedItem<'a>>;

    fn next(&mut self) -> Option<Self::Item> {
        let remaining = &self.input[self.pos..];

        // Skip the first whitespace chars
        let first_non_whitespace_index = remaining.find(|c: char| !c.is_whitespace());
        if remaining.is_empty() || first_non_whitespace_index.is_none() {
            return None;
        }

        let first_non_whitespace_index = first_non_whitespace_index.unwrap();

        let remaining = &remaining[first_non_whitespace_index..];
        self.pos += first_non_whitespace_index;

        let range_start = self.pos;

        // Get the next chars
        let mut chars = remaining.chars().peekable();
        let first_char = chars.next().unwrap();

        if first_char == '"' {
            let (result, read) = parse_string(remaining);
            self.pos += read;
            return Some(RangedObject::new(result, range_start..self.pos));
        } else if first_char == '{' {
            let second = chars.peek();
            if let Some('-') = second {
                let (result, read) = parse_comment(remaining);
                self.pos += read;
                return Some(RangedObject::new(result, range_start..self.pos));
            }
        }

        let (result, read) = parse_block(remaining);
        self.pos += read;
        Some(RangedObject::new(result, range_start..self.pos))
    }
}

fn parse_comment(input: &str) -> (ScannedItem, usize) {
    let chars_pairs = input.chars().zip(input.chars().skip(1));
    let mut index = None;
    for (i, (first, second)) in chars_pairs.enumerate() {
        match (first, second) {
            ('-', '}') => {
                index = Some(i);
                break;
            }
            _ => continue,
        }
    }

    match index {
        Some(i) => (ScannedItem::Comment(&input[2..i]), i + 2),
        None => (ScannedItem::UnclosedComment(&input[2..]), input.len()),
    }
}

fn parse_string(input: &str) -> (ScannedItem, usize) {
    let mut next_is_escape = false;
    let mut index = None;
    for (i, ch) in input.chars().enumerate().skip(1) {
        if !next_is_escape && ch == '\\' {
            next_is_escape = true;
        }

        if !next_is_escape && ch == '"' {
            index = Some(i);
        }

        next_is_escape = false;
    }

    match index {
        Some(i) => (ScannedItem::Str(&input[1..i]), i + 1),
        None => (ScannedItem::UnclosedStr(&input[1..]), input.len()),
    }
}

fn parse_block(input: &str) -> (ScannedItem, usize) {
    let first_whitespace_index = input
        .find(|c: char| c.is_whitespace())
        .unwrap_or_else(|| input.len());

    let first_problematic = input
        .find(|c: char| c == '"' || c == '{')
        .unwrap_or_else(|| input.len());
    let to_read = std::cmp::min(first_whitespace_index, first_problematic);
    (ScannedItem::Rest(&input[0..to_read]), to_read)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn scan_example() {
        let res: Vec<_> = scan("(x <= > >= < y 99.88l8 )").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_example_multiline() {
        let res: Vec<_> = scan("(x <= > >= < y 99.88l8 )\nx ) ( >= y").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_string_escaping() {
        let res: Vec<_> = scan("\"String \\\"inside\\\" a string\"").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_problematic() {
        let res: Vec<_> = scan("x\"something\"").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_comment() {
        let res: Vec<_> = scan("{-some comment-}").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_unclosed_comment() {
        let res: Vec<_> = scan("{-some comment").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_multiline_comment() {
        let res: Vec<_> = scan("{-some comment\n which is long! -}").collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_real_life() {
        let input = "program fib;\r\nbegin\r\nvar n;\r\nvar first := 0;\r\nvar second :=1;\r\nvar next;\r\nvar c :=0 ;\r\nprint \"enter the number of terms\";\r\nget n;\r\nwhile ( c < n)\r\nbegin\r\nif ( c <= 1)\r\nthen begin next := c; end\r\nelse begin\r\n next := first + second;\r\n second := next;\r\nend\r\nprint next;\r\nc := c + 1;\r\nend\r\nend\r\n";
        let res: Vec<_> = scan(input).collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_real_life_with_comments() {
        let input = "program fib;\r\nbegin\r\nvar n;{-a comment here-}\r\nvar first := 0;\r\nvar second :=1;\r\nvar next;\r\nvar c :=0 ;\r\nprint \"enter the number of terms\";\r\nget n;\r\nwhile ( c < n)\r\nbegin\r\nif ( c <= 1)\r\nthen begin next := c; end\r\nelse begin\r\n next := first + second;\r\n second := next;\r\nend\r\nprint next;\r\nc := c + 1;\r\nend\r\nend\r\n";
        let res: Vec<_> = scan(input).collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn scan_oreo_8() {
        let input = r#"
        {- will parse OK -}
program test
                                   
begin                              
                                   
	var n;
	var first := 0;  
	var second :=1; 
	var next;                           
	var c :=0;
	
    print "enter the number of terms";
                 {- comment in body -}
    get n;  
    
    while ( first < (second > 32)  )
    begin

       if ( 42 < 100 ) then begin n := 5;  end
       			   else begin n := 46; end;
                        {- comment
                        over
                        multiple 
                        lines -}
                                  
    c := c + 1;	    
	end;
end    
        "#;
        let res: Vec<_> = scan(input).collect();
        assert_debug_snapshot!(res);
    }
}
