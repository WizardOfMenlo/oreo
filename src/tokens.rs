#[derive(Debug, Clone, PartialEq)]
pub enum Token<'a> {
    Keyword(Keyword),
    Punctuation(Punctuation),
    Operator(Operator),
    Literal(Literal<'a>),
    Identifier(&'a str),
    Comment(&'a str),
    Error(LexicalError<'a>),
}

impl<'a> Token<'a> {
    pub fn is_error(&self) -> bool {
        match self {
            Token::Error(_) => true,
            _ => false,
        }
    }

    pub fn is_comment(&self) -> bool {
        match self {
            Token::Comment(_) => true,
            _ => false,
        }
    }

    pub fn same_kind<'b>(&self, other: &Token<'b>) -> bool
    where
        'b: 'a,
    {
        // This should match every singleton
        if self == other {
            true
        } else {
            // Match literals
            match (self, other) {
                (Self::Literal(Literal::Integer(_)), Self::Literal(Literal::Integer(_))) => true,
                (Self::Literal(Literal::Boolean(_)), Self::Literal(Literal::Boolean(_))) => true,
                (Self::Literal(Literal::String(_)), Self::Literal(Literal::String(_))) => true,
                (Self::Identifier(_), Self::Identifier(_)) => true,
                (Self::Comment(_), Self::Comment(_)) => true,
                _ => false,
            }
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Keyword {
    Program,
    Begin,
    End,
    Var,
    Print,
    Println,
    Get,
    While,
    If,
    Then,
    Else,
    Procedure,
    Return,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Punctuation {
    Comma,
    Semicolon,
    BracketOpen,
    BracketClose,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Operator {
    Assignement,
    Plus,
    Minus,
    Times,
    Divide,
    LesserThan,
    GreaterThan,
    Equals,
    GreaterOrEquals,
    LesserOrEquals,
    Or,
    And,
    Not,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Literal<'a> {
    Integer(isize),
    Boolean(bool),
    String(&'a str),
}

#[derive(Debug, Clone, PartialEq)]
pub enum LexicalError<'a> {
    ExpectedDoubleEqualsEOF,
    ExpectedAssignementEOF,
    UnclosedString(&'a str),
    UnclosedComment(&'a str),
    ExpectedDoubleEquals(char),
    ExpectedAssignement(char),
    UnknownChar(char),
}
