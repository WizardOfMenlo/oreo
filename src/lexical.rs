use crate::scanner::{Scanned, ScannedItem};
use crate::tokens::{Keyword, LexicalError, Literal, Operator, Punctuation, Token};

use lazy_static::lazy_static;
use std::collections::HashMap;

lazy_static! {
    static ref SINGLE_CHAR_TOKEN: HashMap<char, Token<'static>> = {
        let mut m = HashMap::new();
        m.insert('(', Token::Punctuation(Punctuation::BracketOpen));
        m.insert(')', Token::Punctuation(Punctuation::BracketClose));
        m.insert(';', Token::Punctuation(Punctuation::Semicolon));

        m.insert('+', Token::Operator(Operator::Plus));
        m.insert('-', Token::Operator(Operator::Minus));
        m.insert('*', Token::Operator(Operator::Times));
        m.insert('/', Token::Operator(Operator::Divide));

        m
    };
    static ref KEYWORDS: HashMap<&'static str, Token<'static>> = {
        let mut m = HashMap::new();
        m.insert("program", Token::Keyword(Keyword::Program));
        m.insert("begin", Token::Keyword(Keyword::Begin));
        m.insert("end", Token::Keyword(Keyword::End));
        m.insert("var", Token::Keyword(Keyword::Var));
        m.insert("print", Token::Keyword(Keyword::Print));
        m.insert("println", Token::Keyword(Keyword::Println));
        m.insert("get", Token::Keyword(Keyword::Get));
        m.insert("while", Token::Keyword(Keyword::While));
        m.insert("if", Token::Keyword(Keyword::If));
        m.insert("then", Token::Keyword(Keyword::Then));
        m.insert("else", Token::Keyword(Keyword::Else));

        m.insert("or", Token::Operator(Operator::Or));
        m.insert("and", Token::Operator(Operator::And));
        m.insert("not", Token::Operator(Operator::Not));
        m
    };
}

pub fn lexicalize<'a>(it: impl Iterator<Item = Scanned<'a>>) -> impl Iterator<Item = Token<'a>> {
    it.map(lexicalize_one).flatten()
}

fn lexicalize_one<'a>(scan: Scanned<'a>) -> impl Iterator<Item = Token<'a>> {
    LexicalIt::new(scan)
}

struct LexicalIt<'a> {
    s: Scanned<'a>,
    curr_position: usize,
    terminate: bool,
}

impl<'a> LexicalIt<'a> {
    fn new(s: Scanned<'a>) -> Self {
        LexicalIt {
            s,
            curr_position: 0,
            terminate: false,
        }
    }
}

impl<'a> Iterator for LexicalIt<'a> {
    type Item = Token<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.terminate {
            return None;
        }

        let remaining = match self.s.inner {
            ScannedItem::Rest(s) => &s[self.curr_position..],
            ScannedItem::Str(s) => {
                self.terminate = true;
                return Some(Token::Literal(Literal::String(s)));
            }
            ScannedItem::UnclosedStr(s) => {
                self.terminate = true;
                return Some(Token::Error(LexicalError::UnclosedString(s)));
            }
        };

        if remaining.len() == 0 {
            return None;
        }

        let mut chars = remaining.chars();
        let start_char = chars.next().unwrap();

        // This resolves all unambiguos single chars
        if SINGLE_CHAR_TOKEN.contains_key(&start_char) {
            self.curr_position += 1;
            return SINGLE_CHAR_TOKEN.get(&start_char).cloned();
        }

        // Handle the single unambiguos chars
        if start_char == '<' || start_char == '>' || start_char == ':' || start_char == '=' {
            let next = chars.next();
            let (token, move_forward) = match (start_char, next) {
                ('<', Some('=')) => (Token::Operator(Operator::LesserOrEquals), true),
                ('<', None) | ('<', Some(_)) => (Token::Operator(Operator::LesserThan), false),

                ('>', Some('=')) => (Token::Operator(Operator::GreaterOrEquals), true),
                ('>', None) | ('>', Some(_)) => (Token::Operator(Operator::GreaterThan), false),

                (':', Some('=')) => (Token::Operator(Operator::Assignement), true),
                (':', Some(x)) => (Token::Error(LexicalError::ExpectedAssignement(x)), false),
                (':', None) => (Token::Error(LexicalError::ExpectedAssignementEOF), false),

                ('=', Some('=')) => (Token::Operator(Operator::Equals), true),
                ('=', Some(x)) => (Token::Error(LexicalError::ExpectedDoubleEquals(x)), false),
                ('=', None) => (Token::Error(LexicalError::ExpectedDoubleEqualsEOF), false),

                // If we get here, a lot must be wrong
                _ => return Some(Token::Error(LexicalError::UnknownChar('?'))),
            };

            self.curr_position += 1 + move_forward as usize;
            return Some(token);
        }

        // Get a integer literal
        if start_char.is_numeric() {
            let (token, read) = parse_literal(remaining);
            self.curr_position += read;
            return Some(token);
        }

        // Handle stuff that looks like identifiers
        if start_char.is_alphabetic() || start_char == '_' {
            let (token, read) = parse_identifier(remaining);
            self.curr_position += read;
            return Some(token);
        }

        self.curr_position += 1;
        Some(Token::Error(LexicalError::UnknownChar(start_char)))
    }
}

fn parse_literal(input: &str) -> (Token, usize) {
    // ['1', '0', 'a']
    //             ^
    // ['1', '0']
    // 10
    let first_non_number = input
        .find(|c: char| !c.is_numeric())
        .unwrap_or_else(|| input.len());
    let literal = &input[0..first_non_number];
    (
        Token::Literal(Literal::Integer(literal.parse().unwrap())),
        first_non_number,
    )
}

fn parse_identifier(input: &str) -> (Token, usize) {
    let first_non_id = input
        .find(|c: char| !(c.is_alphanumeric() || c == '_'))
        .unwrap_or_else(|| input.len());
    let id = &input[0..first_non_id];

    if KEYWORDS.contains_key(id) {
        return (KEYWORDS.get(id).cloned().unwrap(), first_non_id);
    }

    (Token::Identifier(id), first_non_id)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::scanner::scan;
    use insta::assert_debug_snapshot;

    #[test]
    fn lex_empty() {
        let mut res = lexicalize(scan(""));
        assert_eq!(res.next(), None);
    }

    #[test]
    fn lex_example() {
        let res: Vec<_> = lexicalize(scan("(x <= > >= < y 99.88l8 )")).collect();
        assert_debug_snapshot!(res);
    }

    #[test]
    fn lex_real_life() {
        let input = "program fib;\r\nbegin\r\nvar n;\r\nvar first := 0;\r\nvar second :=1;\r\nvar next;\r\nvar c :=0 ;\r\nprint \"enter the number of terms\";\r\nget n;\r\nwhile ( c < n)\r\nbegin\r\nif ( c <= 1)\r\nthen begin next := c; end\r\nelse begin\r\n next := first + second;\r\n second := next;\r\nend\r\nprint next;\r\nc := c + 1;\r\nend\r\nend\r\n";
        let res: Vec<_> = lexicalize(scan(input)).collect();
        assert_debug_snapshot!(res);
    }
}
