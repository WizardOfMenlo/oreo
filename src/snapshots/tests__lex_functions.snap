---
source: src/lexical.rs
expression: res
---
[
    RangedObject {
        inner: Keyword(
            Procedure,
        ),
        range: 0..9,
    },
    RangedObject {
        inner: Identifier(
            "id",
        ),
        range: 10..12,
    },
    RangedObject {
        inner: Punctuation(
            BracketOpen,
        ),
        range: 12..13,
    },
    RangedObject {
        inner: Keyword(
            Var,
        ),
        range: 13..16,
    },
    RangedObject {
        inner: Identifier(
            "x",
        ),
        range: 17..18,
    },
    RangedObject {
        inner: Punctuation(
            Comma,
        ),
        range: 18..19,
    },
    RangedObject {
        inner: Keyword(
            Var,
        ),
        range: 20..23,
    },
    RangedObject {
        inner: Identifier(
            "y",
        ),
        range: 24..25,
    },
    RangedObject {
        inner: Punctuation(
            BracketClose,
        ),
        range: 25..26,
    },
    RangedObject {
        inner: Keyword(
            Begin,
        ),
        range: 27..32,
    },
    RangedObject {
        inner: Keyword(
            Return,
        ),
        range: 33..39,
    },
    RangedObject {
        inner: Identifier(
            "x",
        ),
        range: 40..41,
    },
    RangedObject {
        inner: Punctuation(
            Semicolon,
        ),
        range: 41..42,
    },
    RangedObject {
        inner: Keyword(
            End,
        ),
        range: 43..46,
    },
]
